### PR TITLE
Feature/cc 16 zoom in and zoom out in design page

### PR DESCRIPTION
**Pull request by ZhihanChenZack**
Add court canvas zoom function.

- The court is zoom base on the center of the canvas as shown.
image

- The maximum zoom-in scale is x1.5 (adjustable zoom scale in the constant file).
- The court is scaled by x1.1 for every click.
- The zoom buttons are disabled at the min/max scale as shown below.

image

- The court will be reset to the original state once switch the blueprints.
- Added the zoom percentage label on the footer as shown.

 image

image

- Konva stage shift calculator is generated in the utils file.

30/9/2022 New update:

- Add a reset button on the editor footer as shown

image

- Add tooltips on all canvas control buttons.

image

If have any questions about the update, contact me :)

### CODE REVIEW

**Review by tatoMa — COMMENTED** [review]
Where is the test of the reducer you created? (zoomCourtSlice)

An example can be found from priceBarSlice.test.ts

File: src/store/reducer/zoomCourtSlice.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  minScale: MIN_ZOOM,
+  maxScale: MAX_ZOOM,
+};
+
+export const ZoomControlSlice = createSlice({
+  name: "ZoomControl",
+  initialState,
+  reducers: {
+    zoomIn: (state: ZoomState) => {
+      state.zoomScale = state.zoomScale * ZOOM_STEP_SIZE;
```
Comment (tatoMa) [inline src/store/reducer/zoomCourtSlice.ts]: Multiple by step or increase by step?

By using multiple, first time is 1 * 1.1 = 1.1
Then 1.1*1.1 = 1.21
Then 1.21*1.1 = 1.331?

or 1 +0.1 = 1.1
then ++0.1 = 1.2 then 1.3, 1.4 and so on.

File: src/components/EditorSideBar/Blueprints.tsx [hunk 2]
```diff
@@ -7,6 +7,7 @@ import { setActiveCourt } from "@/store/reducer/courtSpecDataSlice";
 import { mockTileData } from "../MockData/MockTileData";
 import { useStoreSelector } from "@/store/hooks";
 import { setCartDisplayState } from "@/store/reducer/cartControlSlice";
+import { reset } from "@/store/reducer/zoomCourtSlice";
```
Comment (tatoMa) [inline src/components/EditorSideBar/Blueprints.tsx:10]: The name of this reset can be more meaningful. Cos it is not inside of any zoom related component, devs may not understand what "reset" for.

File: src/store/reducer/zoomCourtSlice.ts (hunk 1 above)
Comment (STY1997) [reply src/store/reducer/zoomCourtSlice.ts]: we changed the algorithm to add/subtract a fixed constance (0.1) in each step.
The min and max scale variable are removed from the slice.

Comment (STY1997) [reply src/store/reducer/zoomCourtSlice.ts]: > And I don't see any Min and Max zoom implementation in this slice. If not used in this slice, they are just constants that are not needed in this slice, or you put them in this slice logic of zoom in and out.

we changed the algorithm to add/subtract a fixed constance (0.1) in each step.
The min and max scale variable are removed from the slice.

File: src/components/EditorSideBar/Blueprints.tsx (hunk 2 above)
Comment (STY1997) [reply src/components/EditorSideBar/Blueprints.tsx]: > The name of this reset can be more meaningful. Cos it is not inside of any zoom related component, devs may not understand what "reset" for.

reset has been renamed as resetZoom

File: src/utils/zoomCenterCalculate.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { IZoomShift } from "@/interfaces/zoomShift";
+
+export const centerZoom = (zoomInfo: IZoomShift) => {
+  const xShift =
```
Comment (ShawnWang86b) [inline src/utils/zoomCenterCalculate.ts:4]: hi Zack and Taylor not understand this formula, can you explain it once you have time
what means yShift, xShift
where is startPoint in the canvas , I found it in console log {X: 2500, Y: 2500}
and dragPos is X=0 and Y=0, where is the initial dragPos(X =0, Y=0) in the canvas
Thanks

File: src/components/FullCourt/index.tsx [hunk 4]
```diff
@@ -11,10 +11,39 @@ import CourtDimension from "../BasketballCourt/CourtDimension";
...
 import useCourt from "@/hooks/useCourt";
+import { IZoomShift } from "@/interfaces/zoomShift";
+import { useStoreSelector } from "@/store/hooks";
+import { centerZoom } from "@/utils/zoomCenterCalculate";
+import { useRef } from "react";
 
 const FullCourt = () => {
   const { courtAreaXLength, courtAreaYLength, borderLength, court, courtStartPoint } = useCourt();
 
+  const zoomScale = useStoreSelector((state) => state.zoomControl.zoomScale);
```
Comment (zhuocun) [inline src/components/FullCourt/index.tsx]: It seems like there are lots of duplicated code in different courts, I'm wondering are there any possible way to make these code reusable? It's totally ok if not. Thank you!

**Review by zhuocun — COMMENTED** [review]
Good job guys, I'v written somthing and hope that can help :)

File: src/components/FullCourt/index.tsx (hunk 4 above)
Comment (ZhihanChenZack) [reply src/components/FullCourt/index.tsx]: New changes related to the duplicate code are updated.

**Review by zhuocun — APPROVED** [review]
Looks really for me :)

**Review by ShawnWang86b — APPROVED** [review]

**Comment by Double-Twelve** [comment]
When I zoom in on the court and drag it, click Download PDF, the PDF generated is the court I dragged, not the original court

image

**Pull request merged**